src/Microsoft.Azure.SignalR.Common/Utilities/BackOffPolicy.cs

```diff
@@ -0,0 +1,112 @@
+            bool probeSuccess = false;
+            do
+            {
+                var myTcs = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
+                // ensure only one caller will be selected to call probing func
+                var ongoingProbeTcs = Interlocked.CompareExchange(ref _currentProbeTcs, myTcs, null);
+                if (ongoingProbeTcs == null)
+                {
+                    // initiate the probe and indicate its result to others
+                    Task<bool> probeTask = null;
+                    bool awaitProbeTask = false;
+                    try
+                    {
+                        Debug.Assert(!calledProbeOnce);
+                        calledProbeOnce = true;
+                        probeTask = probe();
+
+                        using (CancellationTokenSource delayCts = new CancellationTokenSource())
+                        {
+                            var d = getRetryDelay(_currentRetryCount++);
+                            var delayTask = Task.Delay(d, delayCts.Token);
+                            await Task.WhenAny(delayTask, probeTask);
+
+                            // Handle success, timeout, and failure appropriately
+                            if (//probeTask.IsCompletedSuccessfully in .NET Standard 2.1
+                                probeTask.Status == TaskStatus.RanToCompletion && !probeTask.IsFaulted && !probeTask.IsCanceled
+                                && probeTask.Result)
+                            {
+                                // probe is successful
+                                delayCts.Cancel();
+                                probeSuccess = true;
+                            }
+                            else if (delayTask.IsCompleted) //delayTask.IsCompletedSuccessfully
+                            {
+                                // probe timeout
+                                // make sure we still await for the probe task after indicating the failure to others
+                                awaitProbeTask = true;
+                            }
+                            else
+                            {
+                                // probe failed
+                                // make sure we still await for the probe task 
+                                awaitProbeTask = true;
+                                // after waiting for the current backoff time and indicating failure to others
+                                await delayTask;
+                            }
+                        }
+                    }
+                    finally
+                    {
+                        Interlocked.Exchange(ref _currentProbeTcs, null);
+                        // indicate the result to others
+                        myTcs.SetResult(probeSuccess);
+                    }
+
+                    // take care of unfinished probe task
+                    if (awaitProbeTask)
```

Comment:
Should this be inside the lock?

Reply:
It is already protected by https://github.com/Azure/azure-signalr/pull/472/files#diff-bc394dd7d4dd9b3f15a06df56aabd424R38, which plays the role of lock.

Reply:
That's the 'lock' I'm taking about. My question is: in case of probe task timed out, the lock will be released before the probe task finishes, then there will be multiple tasks running at the same time. Is it expected?

Reply:
If the probe task does not succeed (failed or timedout), all the disconnected tasks try to get the 'lock'. The one who fetches the lock starts reconnecting. 
This is a little different from our expected: only 1 disconnected task tries, the others are not involved unless it is successful.
But since the probeTask has risk of hang, if only 1 disconnected task try reconnecting, and it hangs, then the whole reconnecting will hang. So, I think current implementation is Ok.
We should add a cancellation token when StartAsync in https://github.com/Azure/azure-signalr/blob/dev/src/Microsoft.Azure.SignalR.Common/ServiceConnections/ConnectionFactory.cs#L48 to eliminate the potential hang risk.

Reply:
Here is the scenario @chenkennt is likely concerned about:
Connection 1 is chosen as a probe (e.g. enters the 'lock') and the probe hangs. Eventually we time out and release the 'lock' but keep awaiting for the probe's task. At this point another probe can start executing and it will run in parallel with the connection 1 probe. If this is another connection (another index in the array of fixed connections) then this is what we actually want to do: try again without waiting for the slow probe for index 1. The question is: what happens if we initiate another probe for index 1? 

Let's assume this is possible (e.g. someone calls [RestartServiceConnectionCoreAsync](https://github.com/Azure/azure-signalr/blob/8a3c20fca36804290a578bc13c44b95cb5d58df2/src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionContainerBase.cs#L114) for connection 1 while we're awaiting for connection 1 probe). The probe func will create a new connection, replace it in FixedServiceConnections and try to connect. From the service perspective this should not create problems since this is a separate instance. From structural integrity point this is also safe because we already have parallel execution between the retry and un-awaited task from `_ = StartCoreAsync(connection)` call for the same index and the array is protected by the lock. The worst outcome is we "leak" a connection - e.g. if the original probe eventually succeeds it will be unaccounted in the FixedServiceConnections list. In fact this was the reason why it took me so long to create this PR - I made a mistake and instead of awaiting for `awaitProbeTask` task I awaited for a new call to the `probe()` (that is probe became recursive). This worked but it created an excessive number of connections. 

In reality this does not happen because the only place where we probe again is in the finally block of `StartCoreAsync` - that is in the probe func itself.

I agree with @ccic - we should bring the cancellation token (perhaps with some universal timeout). Then the code in `CallProbeWithBackOffAsync` can handle it the same way it handles timeouts.

Reply:
Another concern is - shall we even try to connect multiple connections in parallel? I believe we should. The shorter the disconnected time is, the fewer messages get lost, the better balancing between app servers.

Reply:
I understand even there're multiple connections at the same time their rate is still controlled by the backoff policy. Your explanation makes sense.

Reply:
Sorry I think I misunderstood the original question. If the probe task times out then the `ongoingProbeTcs.Task.Result` is `false` and then whoever was [ awaiting for ongoingProbeTcs.Task](https://github.com/Azure/azure-signalr/blob/8a3c20fca36804290a578bc13c44b95cb5d58df2/src/Microsoft.Azure.SignalR.Common/Utilities/BackOffPolicy.cs#L95) will not call the probe all at once. They will have to select a new shared ongoing probe and await for the new ongoingProbeTcs.Task. It is only in case of a successful probe that everyone else will try to connect at the same time. The thinking is if one has connected successfully then everyone is going to try (network got recovered? service is back?). If the connection attempt failed then we control the rate of retries.

Reply:
This part I already understand. If the probe succeeds all pending probes will connect at the same time. This sounds like a reasonable design.

Reply:
But do we want to control the rate even if it's a successful one? Say thousands of connections connecting at the same time, will it create too much pressure?

Reply:
Successfully establishing a new connection is relatively lightweight. It is not CPU intensive (~2ms of CPU time per connection in my tests), not memory intensive (e.g. no large allocations or spikes during handshake, 5000 connections used 600MB), and not network intensive either. 

Dropping connections could be more expensive (due to contentions while throwing exceptions), but in my quick tests it was still pretty lightweight, about the same as connecting: ~10s of total CPU time for dropping 5000 connections (~2-3 seconds on 4 core server on a local network).  So even creating thousands of them at the same time does not create much pressure.

There could be other consequences of creating too many connections fast (spikes in logging, running out of TCP ports, etc). But these are also the same consequences of simply having too many connections anyway (and these are the reason why app servers want to use the service instead of having clients connecting to them directly). Throttling the rate of connections may help in some cases, but it won’t eliminate the fundamental problem of using too many connections. If in some special case the app server has to have this many connections, I would assume it would rather have them all (re)connected as fast as possible or otherwise it may not serve its purpose.

Reply:
If we ever decide to add some connection rate throttling, the existing code can be adopted to introduce such throttling in 2 small changes:
-	After awaiting for [ongoingProbeTcs.Task](https://github.com/Azure/azure-signalr/pull/472/files#diff-bc394dd7d4dd9b3f15a06df56aabd424R95) one will just continue the loop (instead of calling the probe right away) so its probe func will get queued again along with other callers. This would effectively throttle to 1 concurrent call at a time with the same back off algorithm.
-	Whoever needs a more than 1 concurrent calls can create N instances of BackOffPolicy and round-robin between them. This will allow up to N concurrent connection attempts while maintaining the lock-free nature of the code.

In general case I would assume the number of connections is small and the connection rate throttling isn’t required on app server side.